File: scripts/makecourse/catalog.py
#!/usr/bin/env python3
#
# Copyright (c) 2026, PalEm Dynamics LLC
# Licensed under the Apache License, Version 2.0.
#
"""Load catalog.yaml and resolve shape / runtime / org / extras into a file list.

YAML subset: comments, 2-space indent, mappings, lists of scalars or mappings.
No anchors, tags, or multiline scalars.
"""
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path
from typing import Any


def parse_yaml(text: str) -> Any:
    lines: list[tuple[int, str]] = []
    for raw in text.splitlines():
        stripped = raw.strip()
        if not stripped or stripped.startswith("#"):
            continue
        indent = len(raw) - len(raw.lstrip(" "))
        line = raw.strip()
        if " #" in line and not _is_quoted(line):
            line = line.split(" #", 1)[0].rstrip()
        lines.append((indent, line))
    value, _ = _parse_block(lines, 0, 0)
    return value


def _is_quoted(line: str) -> bool:
    s = line.lstrip("- ").lstrip()
    return s.startswith('"') or s.startswith("'")


def _parse_block(lines: list[tuple[int, str]], i: int, indent: int) -> tuple[Any, int]:
    if i >= len(lines):
        return {}, i
    _, first = lines[i]
    if first.startswith("- "):
        return _parse_list(lines, i, indent)
    return _parse_map(lines, i, indent)


def _parse_map(lines: list[tuple[int, str]], i: int, indent: int) -> tuple[dict[str, Any], int]:
    out: dict[str, Any] = {}
    while i < len(lines):
        ind, line = lines[i]
        if ind < indent:
            break
        if ind > indent:
            raise ValueError(f"bad indent at {line!r}")
        if line.startswith("- "):
            raise ValueError(f"list item where mapping expected: {line!r}")
        key, colon, rest = line.partition(":")
        if not colon:
            raise ValueError(f"expected key: {line!r}")
        key = key.strip()
        rest = rest.strip()
        i += 1
        if rest == "":
            if i < len(lines) and lines[i][0] > indent:
                val, i = _parse_block(lines, i, lines[i][0])
            else:
                val = {}
        else:
            val = _parse_scalar(rest)
        out[key] = val
    return out, i


def _parse_list(lines: list[tuple[int, str]], i: int, indent: int) -> tuple[list[Any], int]:
    out: list[Any] = []
    while i < len(lines):
        ind, line = lines[i]
        if ind < indent:
            break
        if ind > indent:
            raise ValueError(f"bad indent at {line!r}")
        if not line.startswith("- "):
            break
        item = line[2:].strip()
        i += 1
        if item == "":
            if i < len(lines) and lines[i][0] > indent:
                val, i = _parse_block(lines, i, lines[i][0])
            else:
                val = None
        elif ":" in item and not item.startswith(("'", '"')):
            key, _, rest = item.partition(":")
            val = {key.strip(): _parse_scalar(rest.strip()) if rest.strip() else {}}
            while i < len(lines) and lines[i][0] > indent:
                nested, i = _parse_map(lines, i, lines[i][0])
                val.update(nested)
                break
        else:
            val = _parse_scalar(item)
        out.append(val)
    return out, i


def _parse_scalar(text: str) -> Any:
    if text == "" or text in ("null", "~"):
        return None
    if text in ("true", "True"):
        return True
    if text in ("false", "False"):
        return False
    if (text.startswith('"') and text.endswith('"')) or (
        text.startswith("'") and text.endswith("'")
    ):
        return text[1:-1]
    if text.isdigit() or (text.startswith("-") and text[1:].isdigit()):
        return int(text)
    return text


def load_catalog(path: Path) -> dict[str, Any]:
    data = parse_yaml(path.read_text(encoding="utf-8"))
    if not isinstance(data, dict):
        raise ValueError("catalog.yaml must be a mapping")
    return data


def _ids(section: dict[str, Any] | None) -> list[str]:
    if not section:
        return []
    return list(section.keys())


def apply_preset(
    catalog: dict[str, Any],
    *,
    preset: str,
    shape: str,
    runtime: str,
    org: str,
) -> tuple[str, str, str]:
    defaults = catalog.get("defaults") or {}
    if preset:
        presets = catalog.get("presets") or {}
        if preset not in presets:
            known = ", ".join(_ids(presets)) or "(none)"
            raise SystemExit(f"error: unknown --preset {preset!r} (known: {known})")
        spec = presets[preset] or {}
        shape = shape or spec.get("shape") or ""
        runtime = runtime or spec.get("runtime") or ""
        org = org or spec.get("org") or ""
    shape = shape or defaults.get("shape") or "course"
    runtime = runtime or defaults.get("runtime") or "none"
    org = org or defaults.get("org") or "none"
    return shape, runtime, org


def _require(section: dict[str, Any], kind: str, ident: str) -> dict[str, Any]:
    if ident not in section:
        known = ", ".join(section.keys()) or "(none)"
        raise SystemExit(f"error: unknown --{kind} {ident!r} (known: {known})")
    return section[ident] or {}


def _merge_files(layers: list[dict[str, Any]]) -> tuple[list[str], list[dict[str, str]]]:
    dirs: list[str] = []
    seen_dirs: set[str] = set()
    by_dest: dict[str, dict[str, str]] = {}
    for layer in layers:
        for d in layer.get("dirs") or []:
            if d not in seen_dirs:
                seen_dirs.add(d)
                dirs.append(d)
        for item in layer.get("files") or []:
            dest = item["dest"]
            by_dest[dest] = {
                "kind": "file",
                "src": item["src"],
                "dest": dest,
            }
        for item in layer.get("copies") or []:
            dest = item["dest"]
            by_dest[dest] = {
                "kind": "copy",
                "src": item["src"],
                "dest": dest,
                "mode": item.get("mode") or "",
            }
        frag = layer.get("gitignore_append")
        if frag:
            by_dest[f"__append__:{frag}"] = {
                "kind": "append",
                "src": frag,
                "dest": ".gitignore",
            }
        for keep in layer.get("keep") or []:
            by_dest[keep] = {"kind": "keep", "dest": keep}
    return dirs, list(by_dest.values())


def resolve(
    catalog: dict[str, Any],
    *,
    shape: str,
    runtime: str,
    org: str,
    pages: bool,
    migration_docs: bool,
    lms: str,
    github_org: str,
    course_name: str,
) -> dict[str, Any]:
    shapes = catalog.get("shapes") or {}
    runtimes = catalog.get("runtimes") or {}
    orgs = catalog.get("orgs") or {}
    extras = catalog.get("extras") or {}
    defaults = catalog.get("defaults") or {}

    shape_spec = _require(shapes, "shape", shape)
    runtime_spec = _require(runtimes, "runtime", runtime)
    org_spec = _require(orgs, "org", org)

    layers = [shape_spec, runtime_spec, org_spec]
    if pages:
        layers.append(_require(extras, "extra", "pages"))
    if migration_docs:
        layers.append(_require(extras, "extra", "migration_docs"))

    dirs, items = _merge_files(layers)

    vars_: dict[str, str] = {}
    default_lms = str(defaults.get("lms") or "the course LMS")
    vars_["LMS"] = lms or str((org_spec.get("vars") or {}).get("LMS") or default_lms)
    org_github = str((org_spec.get("vars") or {}).get("GITHUB_ORG") or "")
    vars_["GITHUB_ORG"] = github_org or org_github
    for key, value in (org_spec.get("vars") or {}).items():
        if key in ("LMS", "GITHUB_ORG"):
            continue
        vars_[str(key)] = "" if value is None else str(value)

    ghorg = vars_["GITHUB_ORG"]
    if ghorg:
        vars_["CLONE_SNIPPET"] = f"git clone https://github.com/{ghorg}/{course_name}.git"
        vars_["REPO_WEB"] = f"https://github.com/{ghorg}/{course_name}"
        host = vars_.get("GITHUB_PAGES_HOST") or f"{ghorg}.github.io"
        vars_["PAGES_URL"] = f"https://{host}/{course_name}/"
        vars_["GITHUB_REPOSITORY_DEFAULT"] = f"{ghorg}/{course_name}"
    else:
        vars_["CLONE_SNIPPET"] = "git clone <repository-url>"
        vars_["REPO_WEB"] = "the course repository"
        vars_["PAGES_URL"] = "(GitHub Pages URL once enabled)"
        vars_["GITHUB_REPOSITORY_DEFAULT"] = f"YOUR_ORG/{course_name}"
        vars_.setdefault("GITHUB_PAGES_HOST", "")
    vars_.setdefault("FACULTY_DOCS", "")
    vars_.setdefault("GITHUB_PAGES_HOST", vars_.get("GITHUB_PAGES_HOST", ""))

    return {
        "shape": shape,
        "runtime": runtime,
        "org": org,
        "pages": pages,
        "migration_docs": migration_docs,
        "dirs": dirs,
        "items": items,
        "vars": vars_,
    }


def format_list(catalog: dict[str, Any]) -> str:
    lines = [
        f"{catalog.get('kit', {}).get('name', 'symcourse')} catalog",
        "",
        "Defaults:",
        f"  shape    {catalog.get('defaults', {}).get('shape')}",
        f"  runtime  {catalog.get('defaults', {}).get('runtime')}",
        f"  org      {catalog.get('defaults', {}).get('org')}",
        "",
        "Shapes:",
    ]
    for name, spec in (catalog.get("shapes") or {}).items():
        lines.append(f"  {name:12} {(spec or {}).get('description', '')}")
    lines += ["", "Runtimes:"]
    for name, spec in (catalog.get("runtimes") or {}).items():
        lines.append(f"  {name:12} {(spec or {}).get('description', '')}")
    lines += ["", "Orgs:"]
    for name, spec in (catalog.get("orgs") or {}).items():
        lines.append(f"  {name:12} {(spec or {}).get('description', '')}")
    lines += ["", "Presets:"]
    for name, spec in (catalog.get("presets") or {}).items():
        spec = spec or {}
        lines.append(
            f"  {name:12} shape={spec.get('shape')} runtime={spec.get('runtime')} "
            f"org={spec.get('org')}  {(spec.get('description') or '')}"
        )
    lines.append("")
    return "\n".join(lines)


def main(argv: list[str] | None = None) -> int:
    p = argparse.ArgumentParser(prog="catalog.py")
    p.add_argument("--catalog", type=Path, required=True)
    sub = p.add_subparsers(dest="cmd", required=True)

    sub.add_parser("list")

    r = sub.add_parser("resolve")
    r.add_argument("--shape", default="")
    r.add_argument("--runtime", default="")
    r.add_argument("--org", default="")
    r.add_argument("--preset", default="")
    r.add_argument("--pages", action="store_true")
    r.add_argument("--migration-docs", action="store_true")
    r.add_argument("--lms", default="")
    r.add_argument("--github-org", default="")
    r.add_argument("--course-name", default="course")
    r.add_argument(
        "--plain",
        action="store_true",
        help="Emit TAB lines for the shell scaffolder (VAR/DIR/FILE/COPY/APPEND/KEEP)",
    )

    args = p.parse_args(argv)
    catalog = load_catalog(args.catalog)
    if args.cmd == "list":
        sys.stdout.write(format_list(catalog))
        return 0

    shape, runtime, org = apply_preset(
        catalog,
        preset=args.preset,
        shape=args.shape,
        runtime=args.runtime,
        org=args.org,
    )
    resolved = resolve(
        catalog,
        shape=shape,
        runtime=runtime,
        org=org,
        pages=args.pages,
        migration_docs=args.migration_docs,
        lms=args.lms,
        github_org=args.github_org,
        course_name=args.course_name,
    )
    if args.plain:
        sys.stdout.write(_format_plain(resolved))
        return 0
    json.dump(resolved, sys.stdout, indent=2)
    sys.stdout.write("\n")
    return 0


def _format_plain(resolved: dict[str, Any]) -> str:
    lines = [
        f"META\tshape\t{resolved['shape']}",
        f"META\truntime\t{resolved['runtime']}",
        f"META\torg\t{resolved['org']}",
    ]
    for key, value in resolved["vars"].items():
        lines.append(f"VAR\t{key}\t{value}")
    for d in resolved["dirs"]:
        lines.append(f"DIR\t{d}")
    for item in resolved["items"]:
        kind = item["kind"].upper()
        if kind == "KEEP":
            lines.append(f"KEEP\t{item['dest']}")
        elif kind == "APPEND":
            lines.append(f"APPEND\t{item['src']}\t{item['dest']}")
        elif kind == "COPY":
            lines.append(f"COPY\t{item['src']}\t{item['dest']}\t{item.get('mode') or ''}")
        else:
            lines.append(f"FILE\t{item['src']}\t{item['dest']}")
    lines.append("")
    return "\n".join(lines)


if __name__ == "__main__":
    raise SystemExit(main())

File: src/catalog.rs
// Copyright (c) 2026, PalEm Dynamics LLC
// Licensed under the Apache License, Version 2.0.

use std::{
    collections::BTreeMap,
    fs,
    path::Path,
};

use serde::Deserialize;

use crate::error::{
    Error,
    Result,
};

#[derive(Debug, Deserialize)]
struct CatalogFile {
    kit: Kit,
    defaults: Defaults,
    #[serde(default)]
    presets: BTreeMap<String, Preset>,
    #[serde(default)]
    shapes: BTreeMap<String, Layer>,
    #[serde(default)]
    runtimes: BTreeMap<String, Layer>,
    #[serde(default)]
    orgs: BTreeMap<String, Layer>,
    #[serde(default)]
    extras: BTreeMap<String, Layer>,
}

#[derive(Debug, Deserialize)]
struct Kit {
    name: String,
}

#[derive(Debug, Deserialize)]
struct Defaults {
    shape: String,
    runtime: String,
    org: String,
    lms: String,
}

#[derive(Debug, Deserialize, Default)]
struct Preset {
    #[serde(default)]
    description: String,
    #[serde(default)]
    shape: String,
    #[serde(default)]
    runtime: String,
    #[serde(default)]
    org: String,
}

#[derive(Debug, Deserialize, Default)]
struct Layer {
    #[serde(default)]
    description: String,
    #[serde(default)]
    dirs: Vec<String>,
    #[serde(default)]
    keep: Vec<String>,
    #[serde(default)]
    files: Vec<FileSpec>,
    #[serde(default)]
    copies: Vec<CopySpec>,
    #[serde(default)]
    gitignore_append: Option<String>,
    #[serde(default)]
    vars: BTreeMap<String, Option<String>>,
}

#[derive(Debug, Deserialize)]
struct FileSpec {
    src: String,
    dest: String,
}

#[derive(Debug, Deserialize)]
struct CopySpec {
    src: String,
    dest: String,
    #[serde(default)]
    mode: String,
}

#[derive(Debug, Clone)]
pub struct Item {
    pub kind: ItemKind,
    pub src: String,
    pub dest: String,
    pub mode: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ItemKind {
    File,
    Copy,
    Append,
    Keep,
}

impl Item {
    fn key(&self) -> String {
        match self.kind {
            ItemKind::Append => format!("__append__:{}", self.src),
            _ => self.dest.clone(),
        }
    }
}

#[derive(Debug)]
pub struct Resolve<'a> {
    pub shape: &'a str,
    pub runtime: &'a str,
    pub org: &'a str,
    pub pages: bool,
    pub lms: Option<&'a str>,
    pub github_org: Option<&'a str>,
    pub course_name: &'a str,
}

#[derive(Debug)]
pub struct Resolved {
    pub shape: String,
    pub runtime: String,
    pub org: String,
    pub dirs: Vec<String>,
    pub items: Vec<Item>,
    pub vars: BTreeMap<String, String>,
}

#[derive(Debug)]
pub struct Catalog {
    file: CatalogFile,
}

impl Catalog {
    pub fn load(path: &Path) -> Result<Self> {
        let text = fs::read_to_string(path).map_err(|err| {
            if err.kind() == std::io::ErrorKind::NotFound {
                Error::CatalogMissing(path.to_path_buf())
            } else {
                Error::Io(err)
            }
        })?;
        Self::parse(&text)
    }

    pub fn parse(text: &str) -> Result<Self> {
        let file: CatalogFile = serde_yaml::from_str(text)?;
        Ok(Self { file })
    }

    pub fn format_list(&self) -> String {
        let c = &self.file;
        let mut lines = vec![
            format!("{} catalog", c.kit.name),
            String::new(),
            "Defaults:".to_string(),
            format!("  shape    {}", c.defaults.shape),
            format!("  runtime  {}", c.defaults.runtime),
            format!("  org      {}", c.defaults.org),
            String::new(),
            "Shapes:".to_string(),
        ];
        for (name, spec) in &c.shapes {
            lines.push(format!("  {name:<12} {}", spec.description));
        }
        lines.push(String::new());
        lines.push("Runtimes:".to_string());
        for (name, spec) in &c.runtimes {
            lines.push(format!("  {name:<12} {}", spec.description));
        }
        lines.push(String::new());
        lines.push("Orgs:".to_string());
        for (name, spec) in &c.orgs {
            lines.push(format!("  {name:<12} {}", spec.description));
        }
        lines.push(String::new());
        lines.push("Presets:".to_string());
        for (name, spec) in &c.presets {
            lines.push(format!(
                "  {name:<12} shape={} runtime={} org={}  {}",
                spec.shape, spec.runtime, spec.org, spec.description
            ));
        }
        lines.push(String::new());
        lines.join("\n")
    }

    /// Empty strings count as unset, matching the shell scaffolder.
    pub fn apply_preset(
        &self,
        preset: Option<&str>,
        mut shape: Option<&str>,
        mut runtime: Option<&str>,
        mut org: Option<&str>,
    ) -> Result<(String, String, String)> {
        let preset = blank_to_none(preset);
        shape = blank_to_none(shape);
        runtime = blank_to_none(runtime);
        org = blank_to_none(org);

        let mut shape = shape.unwrap_or("").to_string();
        let mut runtime = runtime.unwrap_or("").to_string();
        let mut org = org.unwrap_or("").to_string();

        if let Some(preset) = preset {
            let spec = self.file.presets.get(preset).ok_or_else(|| Error::UnknownId {
                kind: "preset",
                ident: preset.to_string(),
                known: known_keys(&self.file.presets),
            })?;
            if shape.is_empty() {
                shape = spec.shape.clone();
            }
            if runtime.is_empty() {
                runtime = spec.runtime.clone();
            }
            if org.is_empty() {
                org = spec.org.clone();
            }
        }

        if shape.is_empty() {
            shape = self.file.defaults.shape.clone();
        }
        if runtime.is_empty() {
            runtime = self.file.defaults.runtime.clone();
        }
        if org.is_empty() {
            org = self.file.defaults.org.clone();
        }
        Ok((shape, runtime, org))
    }

    pub fn resolve(&self, req: Resolve<'_>) -> Result<Resolved> {
        let shape_spec = require(&self.file.shapes, "shape", req.shape)?;
        let runtime_spec = require(&self.file.runtimes, "runtime", req.runtime)?;
        let org_spec = require(&self.file.orgs, "org", req.org)?;

        let mut layers = vec![shape_spec, runtime_spec, org_spec];
        if req.pages {
            layers.push(require(&self.file.extras, "extra", "pages")?);
        }

        let (dirs, items) = merge_files(&layers);

        let mut vars: BTreeMap<String, String> = BTreeMap::new();
        let default_lms = &self.file.defaults.lms;
        let org_vars = &org_spec.vars;
        let org_lms = var_str(org_vars, "LMS");
        let lms = blank_to_none(req.lms).unwrap_or("");
        vars.insert(
            "LMS".to_string(),
            if lms.is_empty() {
                if org_lms.is_empty() {
                    default_lms.clone()
                } else {
                    org_lms
                }
            } else {
                lms.to_string()
            },
        );
        let org_github = var_str(org_vars, "GITHUB_ORG");
        let github_org = blank_to_none(req.github_org).unwrap_or("");
        let ghorg = if github_org.is_empty() {
            org_github
        } else {
            github_org.to_string()
        };
        vars.insert("GITHUB_ORG".to_string(), ghorg.clone());
        for (key, value) in org_vars {
            if key == "LMS" || key == "GITHUB_ORG" {
                continue;
            }
            vars.insert(key.clone(), value.clone().unwrap_or_default());
        }

        if !ghorg.is_empty() {
            vars.insert(
                "CLONE_SNIPPET".to_string(),
                format!("git clone https://github.com/{ghorg}/{}.git", req.course_name),
            );
            vars.insert(
                "REPO_WEB".to_string(),
                format!("https://github.com/{ghorg}/{}", req.course_name),
            );
            let host = vars
                .get("GITHUB_PAGES_HOST")
                .filter(|h| !h.is_empty())
                .cloned()
                .unwrap_or_else(|| format!("{ghorg}.github.io"));
            vars.insert("PAGES_URL".to_string(), format!("https://{host}/{}/", req.course_name));
            vars.insert(
                "GITHUB_REPOSITORY_DEFAULT".to_string(),
                format!("{ghorg}/{}", req.course_name),
            );
        } else {
            vars.insert("CLONE_SNIPPET".to_string(), "git clone <repository-url>".to_string());
            vars.insert("REPO_WEB".to_string(), "the course repository".to_string());
            vars.insert("PAGES_URL".to_string(), "(GitHub Pages URL once enabled)".to_string());
            vars.insert(
                "GITHUB_REPOSITORY_DEFAULT".to_string(),
                format!("YOUR_ORG/{}", req.course_name),
            );
            vars.entry("GITHUB_PAGES_HOST".to_string()).or_default();
        }
        vars.entry("FACULTY_DOCS".to_string()).or_default();
        vars.entry("GITHUB_PAGES_HOST".to_string()).or_default();

        Ok(Resolved {
            shape: req.shape.to_string(),
            runtime: req.runtime.to_string(),
            org: req.org.to_string(),
            dirs,
            items,
            vars,
        })
    }
}

fn blank_to_none(value: Option<&str>) -> Option<&str> {
    value.filter(|s| !s.is_empty())
}

fn var_str(vars: &BTreeMap<String, Option<String>>, key: &str) -> String {
    vars.get(key).and_then(|v| v.clone()).unwrap_or_default()
}

fn known_keys<T>(map: &BTreeMap<String, T>) -> String {
    if map.is_empty() {
        "(none)".to_string()
    } else {
        map.keys().cloned().collect::<Vec<_>>().join(", ")
    }
}

fn require<'a>(section: &'a BTreeMap<String, Layer>, kind: &'static str, ident: &str) -> Result<&'a Layer> {
    section.get(ident).ok_or_else(|| Error::UnknownId {
        kind,
        ident: ident.to_string(),
        known: known_keys(section),
    })
}

fn merge_files(layers: &[&Layer]) -> (Vec<String>, Vec<Item>) {
    let mut dirs = Vec::new();
    let mut items: Vec<Item> = Vec::new();
    for layer in layers {
        for dir in &layer.dirs {
            if !dirs.iter().any(|d| d == dir) {
                dirs.push(dir.clone());
            }
        }
        for file in &layer.files {
            upsert(
                &mut items,
                Item {
                    kind: ItemKind::File,
                    src: file.src.clone(),
                    dest: file.dest.clone(),
                    mode: String::new(),
                },
            );
        }
        for copy in &layer.copies {
            upsert(
                &mut items,
                Item {
                    kind: ItemKind::Copy,
                    src: copy.src.clone(),
                    dest: copy.dest.clone(),
                    mode: copy.mode.clone(),
                },
            );
        }
        if let Some(frag) = &layer.gitignore_append {
            upsert(
                &mut items,
                Item {
                    kind: ItemKind::Append,
                    src: frag.clone(),
                    dest: ".gitignore".to_string(),
                    mode: String::new(),
                },
            );
        }
        for keep in &layer.keep {
            upsert(
                &mut items,
                Item {
                    kind: ItemKind::Keep,
                    src: String::new(),
                    dest: keep.clone(),
                    mode: String::new(),
                },
            );
        }
    }
    (dirs, items)
}

fn upsert(items: &mut Vec<Item>, item: Item) {
    let key = item.key();
    if let Some(existing) = items.iter_mut().find(|i| i.key() == key) {
        *existing = item;
    } else {
        items.push(item);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn repo_catalog() -> Catalog {
        Catalog::load(&Path::new(env!("CARGO_MANIFEST_DIR")).join("catalog.yaml")).unwrap()
    }

    #[test]
    fn list_names_defaults_and_msia() {
        let text = repo_catalog().format_list();
        assert!(text.contains("shape    course"));
        assert!(text.contains("msia"));
        assert!(text.contains("preset") || text.contains("Presets:"));
    }

    #[test]
    fn preset_msia_fills_empty_flags_and_org_vars() {
        let cat = repo_catalog();
        let (shape, runtime, org) = cat.apply_preset(Some("msia"), None, None, None).unwrap();
        assert_eq!(shape, "course");
        assert_eq!(runtime, "uv");
        assert_eq!(org, "msia");
        let resolved = cat
            .resolve(Resolve {
                shape: &shape,
                runtime: &runtime,
                org: &org,
                pages: false,
                lms: None,
                github_org: None,
                course_name: "ian-6xx",
            })
            .unwrap();
        assert_eq!(resolved.vars["LMS"], "Canvas");
        assert_eq!(resolved.vars["GITHUB_ORG"], "uncg-msia");
        assert!(resolved.vars["CLONE_SNIPPET"].contains("uncg-msia/ian-6xx"));
        assert!(resolved.items.iter().any(|i| i.dest == "Containerfile"));
        assert!(resolved.items.iter().any(|i| i.dest == ".github/workflows/release.yml"));
    }

    #[test]
    fn explicit_flag_overrides_preset() {
        let cat = repo_catalog();
        let (shape, runtime, org) = cat.apply_preset(Some("msia"), None, Some("none"), None).unwrap();
        assert_eq!(runtime, "none");
        assert_eq!(shape, "course");
        assert_eq!(org, "msia");
    }

    #[test]
    fn github_org_override_and_pages() {
        let cat = repo_catalog();
        let resolved = cat
            .resolve(Resolve {
                shape: "course",
                runtime: "none",
                org: "none",
                pages: true,
                lms: Some("Moodle"),
                github_org: Some("example-edu"),
                course_name: "bio-lms",
            })
            .unwrap();
        assert_eq!(resolved.vars["LMS"], "Moodle");
        assert_eq!(
            resolved.vars["CLONE_SNIPPET"],
            "git clone https://github.com/example-edu/bio-lms.git"
        );
        assert_eq!(resolved.vars["PAGES_URL"], "https://example-edu.github.io/bio-lms/");
        assert!(resolved.items.iter().any(|i| i.dest == "site/index.html"));
    }

    #[test]
    fn unknown_shape_names_known_ids() {
        let err = repo_catalog()
            .resolve(Resolve {
                shape: "nope",
                runtime: "none",
                org: "none",
                pages: false,
                lms: None,
                github_org: None,
                course_name: "x",
            })
            .unwrap_err();
        let msg = err.to_string();
        assert!(msg.contains("unknown --shape 'nope'"), "{msg}");
        assert!(msg.contains("course"), "{msg}");
    }
}
